From b0f63936fff0ca76b99e5259d5ae1149d9b3921c Mon Sep 17 00:00:00 2001
From: Magnus Sandin <magnus.sandin@valitron.se>
Date: Fri, 5 Sep 2025 15:23:51 +0200
Subject: [PATCH] Fix build issue using musl reported in #1894

The commit 16fb13d doesn't resolve build issues on Alpine Linux
This patch has been tested with success on Alpine Linux
---
 config.h.in       | 3 +++
 meson.build       | 7 +++++++
 src/collect-io.cc | 6 +++---
 3 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/config.h.in b/config.h.in
index 7f65e8450..74557a53a 100644
--- a/config.h.in
+++ b/config.h.in
@@ -98,6 +98,9 @@
 /* Define to enable lua support */
 #mesondefine HAVE_LUA
 
+/* Define to enable use of mntent.h */
+#mesondefine HAVE_MNTENT_H
+
 /* Define to enable npy support */
 #mesondefine HAVE_NPY
 
diff --git a/meson.build b/meson.build
index 7fcfea741..23fba53cf 100644
--- a/meson.build
+++ b/meson.build
@@ -142,6 +142,13 @@ summary({'gq_appdir': gq_appdir,
 conf_data = configuration_data()
 conf_data.set('DEBUG', debug)
 
+
+# Detect if mntent.h is available
+conf_data.set('HAVE_MNTENT_H', 0)
+if cc.has_header('mntent.h')
+    conf_data.set('HAVE_MNTENT_H', 1)
+endif
+
 conf_data.set('HAVE_GTK4', 0)
 option = get_option('gtk4')
 if option.enabled()
diff --git a/src/collect-io.cc b/src/collect-io.cc
index 7e58e7ae6..d913ed690 100644
--- a/src/collect-io.cc
+++ b/src/collect-io.cc
@@ -29,7 +29,7 @@
 #include <gdk/gdk.h>
 #include <glib-object.h>
 
-#if defined(__GLIBC__)
+#if HAVE_MNTENT_H
 #include <mntent.h>
 #else
 #include <sys/mount.h>
@@ -54,7 +54,7 @@
 namespace
 {
 
-#if defined(__GLIBC__)
+#if HAVE_MNTENT_H
 using MFILE = FILE;
 G_DEFINE_AUTOPTR_CLEANUP_FUNC(MFILE, endmntent)
 #endif
@@ -139,7 +139,7 @@ bool is_file_on_mounted_drive(const gchar *filename)
 		       g_str_has_prefix(filename, dirname);
 	};
 
-#if defined(__GLIBC__)
+#if HAVE_MNTENT_H
 	g_autoptr(MFILE) mount_entries = setmntent("/proc/mounts", "r");
 	if (mount_entries == nullptr)
 		{
